chef/cookbooks/crowbar/recipes/crowbar-upgrade.rb

```diff
@@ -0,0 +1,114 @@
+#
+# Cookbook Name:: crowbar
+# Recipe:: crowbar-upgrade
+#
+# Copyright 2013-2016, SUSE LINUX Products GmbH
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+#
+# This recipe is for executing actions that need to be done at nodes
+# as a preparation for the system upgrade. The preparation itself
+# consists of several steps which we distinguish by various attributes
+# saved in the node structure.
+
+return unless node[:platform] == "suse"
+
+if node["crowbar_wall"]["crowbar_openstack_upgrade"]
+
+  # Actions to be run last, when admin node is already new SUSE Cloud version (6)
+  # Nodes will be restarted and their system upgraded after this.
+
+  # put HA nodes into maintenance mode (TODO is it needed? It wasn't in 4-5 upgrade...)
+
+  # Stop pacemaker resources in all clusters
```

Comment:
Interesting case. First, you don't have access to the sync mark API since you don't depend on the crowbar-pacemaker cookbook and nothing else depends on it. So one solution would be to depend on that cookbook -- but that's relatively bad as we're in the crowbar cookbook here and that shouldn't depend on something like this. Maybe a suggests would work.

Another solution is to actually run what you're running here on all nodes of the cluster, not just the founder. It shouldn't matter much to run on all nodes: it would be important if we were looking at the exit code for crm (because in that case, two crm calls in parallel could possibly conflict with each other). But we do it in a bash resource, so it won't have an impact. The only reason to run on the founder only is to either avoid running the code multiple times (on different nodes) or to guarantee that it's run only once per cluster because the code can create issue. But for "crm resource stop", as mentioned, that should be fine.

But in all cases: yes, you do need to change something because you can't risk stopping the services on some nodes before the pacemaker resources are stopped.